File: client_typescript/search_client.ts
/**
 * Firestore Vector Search Client
 *
 * This module provides TypeScript wrappers for the vector search
 * Cloud Functions, including a React hook for debounced searching.
 *
 * @example
 * ```typescript
 * // Direct function call
 * const results = await searchDocuments({
 *   query: "FinFET process",
 *   limit: 20,
 *   threshold: 0.5
 * });
 *
 * // React hook
 * const { results, isSearching } = useVectorSearch("FinFET process", 0.5);
 * ```
 */

import { getFunctions, httpsCallable, HttpsCallableResult } from 'firebase/functions';
import type {
  SearchRequest,
  SearchResponse,
  SearchResult,
  BackfillRequest,
  BackfillResponse,
  EmbeddingStats,
  FunctionsError,
  isFunctionsError,
} from './types';

// Re-export types for convenience
export type { SearchRequest, SearchResponse, SearchResult, BackfillRequest, BackfillResponse };

// =============================================================================
// Firebase Functions Wrappers
// =============================================================================

/**
 * Search documents using natural language query.
 *
 * Uses vector similarity search to find semantically matching documents.
 *
 * @param request - Search parameters
 * @returns Search results with relevance scores
 * @throws FunctionsError for authentication or validation errors
 *
 * @example
 * ```typescript
 * const response = await searchDocuments({
 *   query: "yield improvement techniques",
 *   limit: 20,
 *   threshold: 0.5
 * });
 *
 * response.results.forEach(result => {
 *   console.log(`${result.relevanceScore}% - ${result.summary}`);
 * });
 * ```
 */
export async function searchDocuments(request: SearchRequest): Promise<SearchResponse> {
  const functions = getFunctions();
  const searchFn = httpsCallable<SearchRequest, SearchResponse>(
    functions,
    'vector_search',
    { timeout: 30000 } // 30 seconds
  );

  const result: HttpsCallableResult<SearchResponse> = await searchFn(request);
  return result.data;
}

/**
 * Backfill embeddings for existing documents.
 *
 * Processes documents that don't have embeddings or have outdated embeddings.
 * Call repeatedly until `remaining` is 0.
 *
 * @param request - Backfill parameters
 * @returns Processing results with counts
 *
 * @example
 * ```typescript
 * // Process in batches until complete
 * let remaining = 1;
 * while (remaining > 0) {
 *   const response = await backfillEmbeddings({ limit: 50 });
 *   console.log(`Processed ${response.processed}, ${response.remaining} remaining`);
 *   remaining = response.remaining;
 * }
 * ```
 */
export async function backfillEmbeddings(request: BackfillRequest = {}): Promise<BackfillResponse> {
  const functions = getFunctions();
  const backfillFn = httpsCallable<BackfillRequest, BackfillResponse>(
    functions,
    'backfill_embeddings',
    { timeout: 540000 } // 9 minutes
  );

  const result = await backfillFn(request);
  return result.data;
}

/**
 * Get embedding coverage statistics.
 *
 * @param collectionPath - Path to the collection to analyze
 * @returns Statistics about embedding coverage
 */
export async function getEmbeddingStats(collectionPath: string = 'documents'): Promise<EmbeddingStats> {
  const functions = getFunctions();
  const statsFn = httpsCallable<{ collectionPath: string }, EmbeddingStats>(
    functions,
    'get_embedding_stats',
    { timeout: 60000 }
  );

  const result = await statsFn({ collectionPath });
  return result.data;
}

// =============================================================================
// React Hook for Debounced Search
// =============================================================================

/**
 * Options for the useVectorSearch hook.
 */
export interface UseVectorSearchOptions {
  /** Debounce delay in milliseconds (default: 500) */
  debounceMs?: number;

  /** Minimum query length to trigger search (default: 3) */
  minQueryLength?: number;

  /** Maximum results to return (default: 20) */
  limit?: number;

  /** Collection to search (default: "documents") */
  collectionPath?: string;
}

/**
 * Return type for useVectorSearch hook.
 */
export interface UseVectorSearchResult {
  /** Search results, null if no search performed */
  results: SearchResult[] | null;

  /** Whether a search is in progress */
  isSearching: boolean;

  /** Error message if search failed */
  error: string | null;

  /** Map of document ID to relevance data for quick lookup */
  resultMap: Map<string, { relevanceScore: number; distance: number }>;
}

/**
 * React hook for debounced vector search.
 *
 * Automatically debounces search queries and manages loading state.
 * Results are cached and only updated when the search completes.
 *
 * @param query - Search query string
 * @param threshold - Distance threshold (0.0-1.0)
 * @param options - Additional options
 * @returns Search results and state
 *
 * @example
 * ```tsx
 * function SearchComponent() {
 *   const [query, setQuery] = useState('');
 *   const [threshold, setThreshold] = useState(0.5);
 *
 *   const { results, isSearching, error } = useVectorSearch(query, threshold);
 *
 *   return (
 *     <div>
 *       <input
 *         value={query}
 *         onChange={(e) => setQuery(e.target.value)}
 *         placeholder="Search documents..."
 *       />
 *       {isSearching && <Spinner />}
 *       {error && <ErrorMessage>{error}</ErrorMessage>}
 *       {results?.map(r => (
 *         <ResultCard key={r.documentId} result={r} />
 *       ))}
 *     </div>
 *   );
 * }
 * ```
 */
export function useVectorSearch(
  query: string,
  threshold: number,
  options: UseVectorSearchOptions = {}
): UseVectorSearchResult {
  // This is a placeholder implementation showing the pattern.
  // In a real app, you'd use React's useState, useEffect, useMemo.

  const {
    debounceMs = 500,
    minQueryLength = 3,
    limit = 20,
    collectionPath = 'documents',
  } = options;

  // Example React implementation (requires React imports):
  /*
  const [results, setResults] = useState<SearchResult[] | null>(null);
  const [isSearching, setIsSearching] = useState(false);
  const [error, setError] = useState<string | null>(null);

  useEffect(() => {
    if (!query.trim() || query.trim().length < minQueryLength) {
      setResults(null);
      setError(null);
      return;
    }

    const timeoutId = setTimeout(async () => {
      setIsSearching(true);
      setError(null);

      try {
        const response = await searchDocuments({
          query,
          limit,
          threshold,
          collectionPath,
        });
        setResults(response.results);
      } catch (err) {
        console.error('Search error:', err);
        setError(err instanceof Error ? err.message : 'Search failed');
        setResults([]);
      } finally {
        setIsSearching(false);
      }
    }, debounceMs);

    return () => clearTimeout(timeoutId);
  }, [query, threshold, limit, collectionPath, debounceMs, minQueryLength]);

  const resultMap = useMemo(() => {
    if (!results) return new Map();
    return new Map(
      results.map(r => [r.documentId, { relevanceScore: r.relevanceScore, distance: r.distance }])
    );
  }, [results]);

  return { results, isSearching, error, resultMap };
  */

  // Placeholder return for documentation purposes
  return {
    results: null,
    isSearching: false,
    error: null,
    resultMap: new Map(),
  };
}

// =============================================================================
// Threshold Persistence
// =============================================================================

const THRESHOLD_STORAGE_KEY = 'vector_search_threshold';

/**
 * Load saved search threshold from localStorage.
 *
 * @param defaultValue - Default threshold if none saved
 * @returns Saved threshold or default
 */
export function loadSearchThreshold(defaultValue: number = 0.5): number {
  if (typeof window === 'undefined') return defaultValue;

  try {
    const saved = localStorage.getItem(THRESHOLD_STORAGE_KEY);
    if (saved) {
      const parsed = parseFloat(saved);
      if (!isNaN(parsed) && parsed >= 0 && parsed <= 1) {
        return parsed;
      }
    }
  } catch {
    // localStorage not available
  }

  return defaultValue;
}

/**
 * Save search threshold to localStorage.
 *
 * @param threshold - Threshold value to save (0.0-1.0)
 */
export function saveSearchThreshold(threshold: number): void {
  if (typeof window === 'undefined') return;

  try {
    localStorage.setItem(THRESHOLD_STORAGE_KEY, threshold.toString());
  } catch {
    // localStorage not available
  }
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Format relevance score for display.
 *
 * @param score - Relevance score (0-100)
 * @param distance - Raw cosine distance
 * @returns Formatted string like "75% (d=0.125)"
 */
export function formatRelevance(score: number, distance: number): string {
  return `${score}% (d=${distance.toFixed(3)})`;
}

/**
 * Check if a document matches search results.
 *
 * @param documentId - Document ID to check
 * @param resultMap - Map from useVectorSearch
 * @returns True if document is in search results
 */
export function isInSearchResults(
  documentId: string,
  resultMap: Map<string, { relevanceScore: number; distance: number }>
): boolean {
  return resultMap.has(documentId);
}

/**
 * Get relevance info for a document from search results.
 *
 * @param documentId - Document ID to look up
 * @param resultMap - Map from useVectorSearch
 * @returns Relevance info or undefined if not in results
 */
export function getRelevanceInfo(
  documentId: string,
  resultMap: Map<string, { relevanceScore: number; distance: number }>
): { relevanceScore: number; distance: number } | undefined {
  return resultMap.get(documentId);
}

File: client_python/search_client.py
"""
Firestore Vector Search Client (Python)

This module provides Python wrappers for the vector search Cloud Functions,
designed for use in Streamlit apps or other Python applications.

Example:
    >>> from search_client import VectorSearchClient
    >>>
    >>> # Initialize with service account
    >>> client = VectorSearchClient("path/to/service-account.json")
    >>>
    >>> # Search documents
    >>> results = client.search("FinFET process with stress engineering")
    >>> for r in results:
    ...     print(f"{r.relevance_score}% - {r.summary}")

Streamlit Example:
    >>> import streamlit as st
    >>> from search_client import VectorSearchClient, StreamlitSearchUI
    >>>
    >>> client = VectorSearchClient.from_streamlit_secrets()
    >>> ui = StreamlitSearchUI(client)
    >>> ui.render()
"""

from __future__ import annotations

import json
import os
import time
from pathlib import Path
from typing import Optional, List, Callable, Any
from functools import lru_cache

import requests

from .types import (
    SearchRequest,
    SearchResponse,
    SearchResult,
    BackfillRequest,
    BackfillResponse,
    EmbeddingStats,
    search_response_from_dict,
    backfill_response_from_dict,
    embedding_stats_from_dict,
)


# =============================================================================
# Vector Search Client
# =============================================================================

class VectorSearchClient:
    """
    Client for interacting with vector search Cloud Functions.

    This client handles authentication and provides typed methods
    for searching documents and managing embeddings.

    Attributes:
        project_id: Firebase project ID.
        region: Cloud Functions region (default: us-central1).

    Example:
        >>> client = VectorSearchClient(
        ...     credentials_path="service-account.json",
        ...     project_id="my-project"
        ... )
        >>>
        >>> results = client.search("metal fill optimization")
        >>> print(f"Found {len(results.results)} matches")
    """

    def __init__(
        self,
        credentials_path: Optional[str] = None,
        project_id: Optional[str] = None,
        region: str = "us-central1",
        id_token: Optional[str] = None,
    ):
        """
        Initialize the vector search client.

        Args:
            credentials_path: Path to service account JSON file.
            project_id: Firebase project ID (inferred from credentials if not provided).
            region: Cloud Functions region.
            id_token: Pre-obtained ID token for authentication (for Streamlit/web apps).
        """
        self.region = region
        self._id_token = id_token
        self._credentials_path = credentials_path

        # Initialize Firebase Admin if credentials provided
        if credentials_path:
            self._init_firebase_admin(credentials_path)

            # Infer project ID from credentials if not provided
            if not project_id:
                with open(credentials_path) as f:
                    creds = json.load(f)
                    project_id = creds.get("project_id")

        if not project_id:
            project_id = os.environ.get("GOOGLE_CLOUD_PROJECT")

        if not project_id:
            raise ValueError(
                "project_id must be provided or set via GOOGLE_CLOUD_PROJECT environment variable"
            )

        self.project_id = project_id
        self._base_url = f"https://{region}-{project_id}.cloudfunctions.net"

    def _init_firebase_admin(self, credentials_path: str) -> None:
        """Initialize Firebase Admin SDK."""
        try:
            import firebase_admin
            from firebase_admin import credentials

            # Only initialize if not already done
            if not firebase_admin._apps:
                cred = credentials.Certificate(credentials_path)
                firebase_admin.initialize_app(cred)
        except ImportError:
            # firebase_admin not installed, will use REST API
            pass

    @classmethod
    def from_streamlit_secrets(cls, region: str = "us-central1") -> "VectorSearchClient":
        """
        Create client from Streamlit secrets.

        Expects secrets in .streamlit/secrets.toml:

        [firebase]
        project_id = "your-project"

        [firebase.credentials]
        type = "service_account"
        project_id = "your-project"
        ...

        Args:
            region: Cloud Functions region.

        Returns:
            Configured VectorSearchClient.
        """
        try:
            import streamlit as st

            # Get credentials from secrets
            creds = dict(st.secrets.get("firebase", {}).get("credentials", {}))
            project_id = st.secrets.get("firebase", {}).get("project_id")

            if not creds or not project_id:
                raise ValueError(
                    "Firebase credentials not found in Streamlit secrets. "
                    "Add [firebase] section to .streamlit/secrets.toml"
                )

            # Write credentials to temp file for Firebase Admin
            import tempfile
            with tempfile.NamedTemporaryFile(mode='w', suffix='.json', delete=False) as f:
                json.dump(creds, f)
                creds_path = f.name

            return cls(
                credentials_path=creds_path,
                project_id=project_id,
                region=region,
            )

        except ImportError:
            raise ImportError("streamlit package required for from_streamlit_secrets()")

    @classmethod
    def from_environment(cls, region: str = "us-central1") -> "VectorSearchClient":
        """
        Create client from environment variables.

        Expects:
        - GOOGLE_APPLICATION_CREDENTIALS: Path to service account JSON
        - GOOGLE_CLOUD_PROJECT: Firebase project ID

        Args:
            region: Cloud Functions region.

        Returns:
            Configured VectorSearchClient.
        """
        creds_path = os.environ.get("GOOGLE_APPLICATION_CREDENTIALS")
        project_id = os.environ.get("GOOGLE_CLOUD_PROJECT")

        return cls(
            credentials_path=creds_path,
            project_id=project_id,
            region=region,
        )

    def _get_id_token(self) -> str:
        """Get ID token for authenticating Cloud Function calls."""
        if self._id_token:
            return self._id_token

        try:
            import google.auth
            from google.auth.transport.requests import Request
            from google.oauth2 import id_token

            # Get credentials
            credentials, _ = google.auth.default()

            # Refresh if needed
            if hasattr(credentials, 'refresh'):
                credentials.refresh(Request())

            # Get ID token for the Cloud Function
            target_audience = f"{self._base_url}/vector_search"
            token = id_token.fetch_id_token(Request(), target_audience)

            return token

        except Exception as e:
            raise RuntimeError(f"Failed to get ID token: {e}")

    def _call_function(self, function_name: str, data: dict) -> dict:
        """
        Call a Cloud Function via HTTP.

        Args:
            function_name: Name of the function to call.
            data: Request data to send.

        Returns:
            Response data from the function.

        Raises:
            RuntimeError: If the function call fails.
        """
        url = f"{self._base_url}/{function_name}"

        headers = {
            "Content-Type": "application/json",
        }

        # Add authentication if we have credentials
        if self._credentials_path or self._id_token:
            try:
                token = self._get_id_token()
                headers["Authorization"] = f"Bearer {token}"
            except Exception:
                pass  # Try without auth (for testing)

        response = requests.post(
            url,
            json={"data": data},
            headers=headers,
            timeout=60,
        )

        if response.status_code != 200:
            raise RuntimeError(
                f"Function {function_name} failed with status {response.status_code}: "
                f"{response.text}"
            )

        result = response.json()

        # Cloud Functions wrap response in {"result": ...}
        if "result" in result:
            return result["result"]

        return result

    def search(
        self,
        query: str,
        limit: int = 20,
        threshold: float = 0.5,
        collection_path: str = "documents",
    ) -> SearchResponse:
        """
        Search documents using natural language query.

        Args:
            query: Natural language search query (min 3 characters).
            limit: Maximum results to return (default: 20, max: 50).
            threshold: Distance threshold 0.0-1.0 (default: 0.5).
            collection_path: Collection to search (default: "documents").

        Returns:
            SearchResponse with results sorted by relevance.

        Raises:
            ValueError: If query is too short.
            RuntimeError: If search fails.

        Example:
            >>> response = client.search("yield improvement techniques")
            >>> for result in response.results:
            ...     print(f"{result.relevance_score}% - {result.summary}")
        """
        if len(query.strip()) < 3:
            raise ValueError("Query must be at least 3 characters")

        data = {
            "collectionPath": collection_path,
            "query": query,
            "limit": min(limit, 50),
            "threshold": max(0.0, min(1.0, threshold)),
        }

        result = self._call_function("vector_search", data)
        return search_response_from_dict(result)

    def backfill_embeddings(
        self,
        collection_path: str = "documents",
        limit: int = 50,
    ) -> BackfillResponse:
        """
        Backfill embeddings for existing documents.

        Call repeatedly until remaining is 0.

        Args:
            collection_path: Collection to process.
            limit: Maximum documents per call (default: 50, max: 200).

        Returns:
            BackfillResponse with processing counts.

        Example:
            >>> while True:
            ...     response = client.backfill_embeddings(limit=50)
            ...     print(f"Processed {response.processed}, {response.remaining} left")
            ...     if response.remaining == 0:
            ...         break
        """
        data = {
            "collectionPath": collection_path,
            "limit": min(limit, 200),
        }

        result = self._call_function("backfill_embeddings", data)
        return backfill_response_from_dict(result)

    def get_embedding_stats(
        self,
        collection_path: str = "documents",
    ) -> EmbeddingStats:
        """
        Get embedding coverage statistics.

        Args:
            collection_path: Collection to analyze.

        Returns:
            EmbeddingStats with coverage information.

        Example:
            >>> stats = client.get_embedding_stats()
            >>> print(f"Coverage: {stats.coverage_percent}%")
        """
        data = {"collectionPath": collection_path}
        result = self._call_function("get_embedding_stats", data)
        return embedding_stats_from_dict(result)


# =============================================================================
# Streamlit UI Components
# =============================================================================

class StreamlitSearchUI:
    """
    Ready-to-use Streamlit UI for vector search.

    Provides a complete search interface with debouncing,
    threshold configuration, and result display.

    Example:
        >>> import streamlit as st
        >>> from search_client import VectorSearchClient, StreamlitSearchUI
        >>>
        >>> client = VectorSearchClient.from_streamlit_secrets()
        >>> ui = StreamlitSearchUI(client)
        >>> ui.render()
    """

    def __init__(
        self,
        client: VectorSearchClient,
        collection_path: str = "documents",
        debounce_ms: int = 500,
        default_threshold: float = 0.5,
    ):
        """
        Initialize the Streamlit search UI.

        Args:
            client: VectorSearchClient instance.
            collection_path: Default collection to search.
            debounce_ms: Debounce delay for search input.
            default_threshold: Default search threshold.
        """
        self.client = client
        self.collection_path = collection_path
        self.debounce_ms = debounce_ms
        self.default_threshold = default_threshold

    def render(self) -> Optional[SearchResponse]:
        """
        Render the search UI and return results.

        Returns:
            SearchResponse if search was performed, None otherwise.
        """
        import streamlit as st

        # Initialize session state
        if "search_query" not in st.session_state:
            st.session_state.search_query = ""
        if "search_threshold" not in st.session_state:
            st.session_state.search_threshold = self.default_threshold
        if "search_results" not in st.session_state:
            st.session_state.search_results = None

        # Search input
        query = st.text_input(
            "Search",
            value=st.session_state.search_query,
            placeholder="Enter search query (min 3 characters)...",
            key="search_input",
        )

        # Threshold slider in expander
        with st.expander("Search Settings"):
            threshold = st.slider(
                "Distance Threshold",
                min_value=0.0,
                max_value=1.0,
                value=st.session_state.search_threshold,
                step=0.05,
                help="Lower = stricter matching. Higher = more results.",
            )
            st.session_state.search_threshold = threshold

            # Show similarity interpretation
            similarity = int((1 - threshold / 2) * 100)
            st.caption(f"Showing results at least {similarity}% similar")

        # Perform search
        results = None
        if query and len(query.strip()) >= 3:
            if query != st.session_state.search_query:
                st.session_state.search_query = query

                with st.spinner("Searching..."):
                    try:
                        results = self.client.search(
                            query=query,
                            threshold=threshold,
                            collection_path=self.collection_path,
                        )
                        st.session_state.search_results = results
                    except Exception as e:
                        st.error(f"Search failed: {e}")
                        return None
            else:
                results = st.session_state.search_results

        # Display results
        if results:
            st.subheader(f"Results ({len(results.results)} matches)")

            for result in results.results:
                self._render_result(result)
        elif query and len(query.strip()) >= 3:
            st.info("No matching documents found.")

        return results

    def _render_result(self, result: SearchResult) -> None:
        """Render a single search result."""
        import streamlit as st

        # Relevance badge color
        if result.relevance_score >= 70:
            badge_color = "green"
        elif result.relevance_score >= 40:
            badge_color = "orange"
        else:
            badge_color = "gray"

        with st.container():
            col1, col2 = st.columns([1, 4])

            with col1:
                st.markdown(
                    f"<span style='color: {badge_color}; font-size: 1.5em; font-weight: bold;'>"
                    f"{result.relevance_score}%</span>",
                    unsafe_allow_html=True,
                )
                st.caption(f"d={result.distance:.3f}")

            with col2:
                if result.title:
                    st.markdown(f"**{result.title}**")
                st.write(result.summary)
                if result.category:
                    st.caption(f"Category: {result.category}")

            st.divider()


# =============================================================================
# Utility Functions
# =============================================================================

def calculate_relevance(distance: float, baseline: float = 0.4) -> int:
    """
    Convert cosine distance to relevance percentage.

    Args:
        distance: Cosine distance (0.0-2.0).
        baseline: Distance at which relevance becomes 0% (default: 0.4).

    Returns:
        Relevance score 0-100.

    Example:
        >>> calculate_relevance(0.1)
        75
        >>> calculate_relevance(0.4)
        0
    """
    relevance = (1 - distance / baseline) * 100
    return max(0, min(100, int(relevance)))


def format_relevance(score: int, distance: float) -> str:
    """
    Format relevance for display.

    Args:
        score: Relevance score (0-100).
        distance: Raw cosine distance.

    Returns:
        Formatted string like "75% (d=0.100)".
    """
    return f"{score}% (d={distance:.3f})"


def batch_backfill(
    client: VectorSearchClient,
    collection_path: str = "documents",
    batch_size: int = 50,
    progress_callback: Optional[Callable[[int, int], None]] = None,
) -> int:
    """
    Run backfill until all documents have embeddings.

    Args:
        client: VectorSearchClient instance.
        collection_path: Collection to process.
        batch_size: Documents per batch.
        progress_callback: Optional callback(processed, remaining).

    Returns:
        Total documents processed.

    Example:
        >>> def on_progress(processed, remaining):
        ...     print(f"Processed {processed}, {remaining} remaining")
        >>>
        >>> total = batch_backfill(client, progress_callback=on_progress)
        >>> print(f"Backfilled {total} documents")
    """
    total_processed = 0

    while True:
        response = client.backfill_embeddings(
            collection_path=collection_path,
            limit=batch_size,
        )

        total_processed += response.processed

        if progress_callback:
            progress_callback(total_processed, response.remaining)

        if response.remaining == 0:
            break

        # Brief pause to avoid rate limiting
        time.sleep(0.5)

    return total_processed
